File: src/db/genre-canonical.js
// Genre name canonicalisation using the MusicBrainz reference list.
//
// The DB problem this solves: pre-V35, the scanner did
//   SELECT id FROM genres WHERE name = ?
// which is case-sensitive AND punctuation-sensitive. A library with
// files tagged "Hip-Hop", "Hip Hop", and "hip-hop" would create three
// distinct rows in `genres`, three distinct dropdown entries in the
// Auto-DJ panel, and three distinct piles when the user wanted to
// filter for "Hip Hop tracks."
//
// V34 made every READ-side lookup `COLLATE NOCASE`, which fixed the
// case-difference subset (Jazz/jazz now resolve to one row at query
// time). But two rows still exist in `genres`, and punctuation
// variants (Hip-Hop vs Hip Hop) still don't merge.
//
// V35 closes both gaps by canonicalising at the WRITE side:
//   1. Normalise the input string (case, separators, ampersand).
//   2. Apply hand-curated display overrides (acronyms etc.) first.
//   3. Otherwise look up against the bundled MusicBrainz genre list
//      (2,141 entries, CC0 — see data/mb-genres.json).
//   4. If MB recognises it, return the Title Case form.
//   5. If unknown to MB, preserve the user's typing.
//
// All callers should go through findOrCreateGenre in manager.js,
// which calls canonicalGenreName before looking up / inserting.
//
// ── Updating the list ────────────────────────────────────────────
// Run `node scripts/refresh-mb-genres.js` to refresh
// data/mb-genres.json from the MB API. The list grows ~monthly;
// quarterly refreshes are plenty.

import fs from 'node:fs';
import path from 'node:path';
import { getDirname } from '../util/esm-helpers.js';

const __dirname = getDirname(import.meta.url);
const MB_JSON_PATH = path.resolve(__dirname, '..', '..', 'data', 'mb-genres.json');

// Normalise a raw genre string for lookup. Lowercase, separators
// (hyphen / underscore) collapsed to spaces, `&` folded to " and ",
// whitespace collapsed, trimmed. Idempotent.
//
// Both the MB list (at load time) and scanner input (at lookup time)
// pass through this same function, so dupes collapse identically
// regardless of which side has hyphens / ampersands / mixed case.
//
// Examples:
//   "Hip-Hop"      → "hip hop"
//   "Hip Hop"      → "hip hop"
//   "K-Pop"        → "k pop"
//   "Drum & Bass"  → "drum and bass"
//   "R&B"          → "r and b"
//   "  RnB  "      → "rnb"
//
// Exported for tests + the V35 migration's use.
export function normaliseForLookup(s) {
  if (typeof s !== 'string') { return ''; }
  return s
    .toLowerCase()
    .replace(/[-_]/g, ' ')
    .replace(/\s*&\s*/g, ' and ')
    .replace(/\s+/g, ' ')
    .trim();
}

// Title-case a normalised genre name for display. Capitalises the
// first letter of each whitespace- or hyphen-delimited word, leaves
// the rest as-is. Doesn't handle acronyms — those need to be in
// DISPLAY_OVERRIDES because naive title-case gives "Edm" / "Idm" /
// "R&b". The override layer is the safety net.
function titleCase(s) {
  return s.replace(/(^|[\s-])([a-z])/g, (_, sep, ch) => sep + ch.toUpperCase());
}

// Display-form overrides — keyed by the NORMALISED form (post-
// `normaliseForLookup`). Looked up BEFORE the MB set so we can
// override the display even for entries MB recognises (e.g. MB has
// "k-pop" which normalises to "k pop"; we want to display "K-Pop",
// not "K Pop"). May also include entries that aren't in MB if we
// have a strong opinion (e.g. "rnb" → "R&B" lets users who tag
// without the ampersand still land on the canonical R&B name).
//
// Hand-curated. To add an entry: pick a genre, work out its
// normalised key via `normaliseForLookup`, decide on a display
// form. Keep this list short — Title Case + MB lookup covers most
// of the surface; the override map is for cases where Title Case
// is visibly wrong.
const DISPLAY_OVERRIDES = new Map([
  ['edm',                'EDM'],
  ['idm',                'IDM'],
  ['ebm',                'EBM'],
  ['mpb',                'MPB'],
  ['r and b',            'R&B'],
  ['rnb',                'R&B'],
  ['k pop',              'K-Pop'],
  ['j pop',              'J-Pop'],
  ['j rock',             'J-Rock'],
  ['c pop',              'C-Pop'],
  ['lo fi',              'Lo-Fi'],
  ['nu metal',           'Nu Metal'],
  ['nu jazz',            'Nu Jazz'],
  ['uk garage',          'UK Garage'],
  ['uk drill',           'UK Drill'],
  ['uk hip hop',         'UK Hip Hop'],
  ['us power metal',     'US Power Metal'],
  ['g funk',             'G-Funk'],
  ['nu disco',           'Nu-Disco'],
  ['edm trap',           'EDM Trap'],
]);

// Loaded once at module init. The JSON is small (~40KB), parsed eagerly
// so the scanner's hot path never blocks on I/O.
//
// Each entry is run through `normaliseForLookup` so the Set holds the
// post-normalisation form. That way input → normalise → MB-set lookup
// converges regardless of separator style.
//
// Wrapped in try/catch so a missing data file (e.g. a half-checked-out
// repo, a Docker layer that didn't ship `data/`) doesn't crash boot —
// canonicalisation degrades to "preserve user input" and the system
// still works, just without dedup.
let MB_GENRES = new Set();
try {
  const raw = fs.readFileSync(MB_JSON_PATH, 'utf8');
  const parsed = JSON.parse(raw);
  if (Array.isArray(parsed.genres)) {
    MB_GENRES = new Set(parsed.genres.map(g => normaliseForLookup(String(g))));
  }
} catch (_) {
  // Don't import winston here — this module loads before logger is
  // configured. Silently degrade. The data file's presence is verified
  // by the test suite.
}

// Returns the canonical display form of a raw genre tag, or null
// for empty input.
//
// Examples (assuming MB has "hip hop", "edm", "drum and bass"):
//   "Hip-Hop"       → "Hip Hop"     (MB match, title-cased)
//   "HIP HOP"       → "Hip Hop"     (MB match, title-cased)
//   "hip hop"       → "Hip Hop"     (MB match, title-cased)
//   "EDM"           → "EDM"         (override)
//   "edm"           → "EDM"         (override)
//   "K-Pop"         → "K-Pop"       (override)
//   "k-pop"         → "K-Pop"       (override)
//   "Drum & Bass"   → "Drum and Bass" (MB match — "drum and bass" — title-cased)
//   "R&B"           → "R&B"         (override)
//   "RnB"           → "R&B"         (override)
//   "Vaporwave"     → "Vaporwave"   (MB match, title-cased)
//   "VaporTwitch"   → "VaporTwitch" (unknown → preserve user input)
//   "  "            → null         (empty)
//   null            → null
export function canonicalGenreName(rawTag) {
  if (rawTag == null) { return null; }
  const trimmed = String(rawTag).trim();
  if (!trimmed) { return null; }
  const normalised = normaliseForLookup(trimmed);
  if (!normalised) { return null; }

  // Override first — keyed on the normalised form, applies regardless
  // of whether MB also has the entry. Lets us provide a preferred
  // display form for things like "k-pop" (MB has it but we want
  // "K-Pop" not "K Pop") AND things MB doesn't have (e.g. "RnB" → "R&B").
  if (DISPLAY_OVERRIDES.has(normalised)) {
    return DISPLAY_OVERRIDES.get(normalised);
  }

  // Known to MB — title-case the normalised form for display.
  if (MB_GENRES.has(normalised)) {
    return titleCase(normalised);
  }

  // Unknown — return the user's input unchanged. Letting a typo or
  // niche genre go through with its original casing is better than
  // mangling it via title-case (which would still be wrong for
  // acronyms we haven't catalogued).
  return trimmed;
}

// Test-only export so tests can assert the bundle was loaded
// correctly without re-reading the JSON file.
export const _internals = {
  mbGenresSize: () => MB_GENRES.size,
  hasMbGenre: (s) => MB_GENRES.has(s),
};

File: rust-parser/src/genre_canonical.rs
// Genre name canonicalisation — Rust mirror of src/db/genre-canonical.js.
//
// Both scanners (Rust prebuilt + JS fallback) call this on every
// genre-tag string before looking up / inserting in the `genres` table.
// V35 onwards: case + punctuation dupes get collapsed at scan time
// (e.g. "Hip-Hop", "Hip Hop", "hip hop" all → "Hip Hop"), so the
// genres table holds one row per canonical concept rather than one per
// user typing.
//
// The MB list (data/mb-genres.json — CC0) is embedded at compile time
// via include_str! so the binary is self-contained. Refresh via
// `node scripts/refresh-mb-genres.js`, commit the updated JSON, then
// rebuild the Rust binary (CI handles the rebuild on push, matches the
// existing pattern for any rust-parser source change).
//
// If the embedded JSON ever fails to parse (shouldn't happen — the
// refresh script writes a stable schema, the test suite verifies the
// shape), we degrade to "preserve user input" and the system still
// works, just without the dedup benefit. That parity with the JS
// module's silent-degrade behaviour is intentional.

use std::collections::{HashMap, HashSet};
use std::sync::OnceLock;

use serde::Deserialize;

#[derive(Deserialize)]
struct MbGenresFile {
    genres: Vec<String>,
}

// JSON is baked into the binary at compile time. The path is relative
// to this source file: rust-parser/src/genre_canonical.rs → ../../data/mb-genres.json.
const MB_GENRES_JSON: &str = include_str!("../../data/mb-genres.json");

// Normalise a raw genre string for lookup. Mirrors the JS
// `normaliseForLookup` function exactly: lowercase, hyphen/underscore
// → space, `&` → " and ", whitespace collapsed, trimmed. Idempotent.
// Both the MB list (at load time) and scanner input (at lookup time)
// pass through this same function so dupes collapse identically.
fn normalise_for_lookup(s: &str) -> String {
    let lowercased: String = s.to_lowercase();
    // First pass: char replacements (`&` → " and ", `-` / `_` → space).
    let mut expanded = String::with_capacity(lowercased.len() + 8);
    for c in lowercased.chars() {
        match c {
            '&' => expanded.push_str(" and "),
            '-' | '_' => expanded.push(' '),
            other => expanded.push(other),
        }
    }
    // Second pass: collapse runs of whitespace + trim.
    let mut out = String::with_capacity(expanded.len());
    let mut prev_space = false;
    for c in expanded.chars() {
        if c.is_whitespace() {
            if !prev_space && !out.is_empty() {
                out.push(' ');
                prev_space = true;
            }
        } else {
            out.push(c);
            prev_space = false;
        }
    }
    if out.ends_with(' ') {
        out.pop();
    }
    out
}

fn mb_genres() -> &'static HashSet<String> {
    static CELL: OnceLock<HashSet<String>> = OnceLock::new();
    CELL.get_or_init(|| match serde_json::from_str::<MbGenresFile>(MB_GENRES_JSON) {
        Ok(parsed) => parsed
            .genres
            .into_iter()
            .map(|g| normalise_for_lookup(&g))
            .collect(),
        // Silent degrade — same as the JS module. The test suite
        // verifies the file parses; if we ever hit this branch in
        // production we still produce correct (just non-deduped) data.
        Err(_) => HashSet::new(),
    })
}

// Display-form overrides for genres whose Title Case form would be
// wrong (acronyms, special spellings). Keyed by the NORMALISED form
// (post-`normalise_for_lookup`). Looked up BEFORE the MB set —
// applies even for entries MB recognises (MB has "k-pop" which
// normalises to "k pop"; we want display "K-Pop", not "K Pop").
//
// Must stay in lock-step with the JS module's DISPLAY_OVERRIDES —
// there's a parity test that verifies both sides produce the same
// output for a fixed input set.
fn display_overrides() -> &'static HashMap<&'static str, &'static str> {
    static CELL: OnceLock<HashMap<&'static str, &'static str>> = OnceLock::new();
    CELL.get_or_init(|| {
        let mut m = HashMap::new();
        m.insert("edm", "EDM");
        m.insert("idm", "IDM");
        m.insert("ebm", "EBM");
        m.insert("mpb", "MPB");
        m.insert("r and b", "R&B");
        m.insert("rnb", "R&B");
        m.insert("k pop", "K-Pop");
        m.insert("j pop", "J-Pop");
        m.insert("j rock", "J-Rock");
        m.insert("c pop", "C-Pop");
        m.insert("lo fi", "Lo-Fi");
        m.insert("nu metal", "Nu Metal");
        m.insert("nu jazz", "Nu Jazz");
        m.insert("uk garage", "UK Garage");
        m.insert("uk drill", "UK Drill");
        m.insert("uk hip hop", "UK Hip Hop");
        m.insert("us power metal", "US Power Metal");
        m.insert("g funk", "G-Funk");
        m.insert("nu disco", "Nu-Disco");
        m.insert("edm trap", "EDM Trap");
        m
    })
}

// Title-case a normalised genre name. Capitalise the first letter of
// each whitespace- or hyphen-delimited word. Mirrors the JS version.
fn title_case(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    let mut capitalise_next = true;
    for c in s.chars() {
        if c.is_whitespace() || c == '-' {
            out.push(c);
            capitalise_next = true;
        } else if capitalise_next {
            out.extend(c.to_uppercase());
            capitalise_next = false;
        } else {
            out.push(c);
        }
    }
    out
}

/// Returns the canonical display form of a raw genre tag.
/// - `None` if input is empty / whitespace-only.
/// - `Some(override)` if the normalised form has a display override.
/// - `Some(title_case)` if MB recognises the normalised form.
/// - `Some(trimmed)` if MB doesn't recognise it, preserving user's casing.
///
/// Mirrors `canonicalGenreName` in src/db/genre-canonical.js — parity
/// is required because both scanners write to the same genres table.
pub fn canonical_genre_name(raw: &str) -> Option<String> {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return None;
    }
    let normalised = normalise_for_lookup(trimmed);
    if normalised.is_empty() {
        return None;
    }

    // Override map first — applies regardless of MB presence.
    if let Some(display) = display_overrides().get(normalised.as_str()) {
        return Some((*display).to_string());
    }

    // MB known — title-case for display.
    if mb_genres().contains(&normalised) {
        return Some(title_case(&normalised));
    }

    // Unknown — preserve user's casing.
    Some(trimmed.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalise_handles_basic_cases() {
        assert_eq!(normalise_for_lookup("Hip-Hop"), "hip hop");
        assert_eq!(normalise_for_lookup("  Hip   Hop  "), "hip hop");
        assert_eq!(normalise_for_lookup("Drum & Bass"), "drum and bass");
        assert_eq!(normalise_for_lookup("R&B"), "r and b");
        assert_eq!(normalise_for_lookup("K-Pop"), "k pop");
    }

    #[test]
    fn title_case_capitalises_word_boundaries() {
        assert_eq!(title_case("hip hop"), "Hip Hop");
        assert_eq!(title_case("lo-fi"), "Lo-Fi");
        assert_eq!(title_case("drum and bass"), "Drum And Bass");
    }

    #[test]
    fn canonical_matches_mb_entries() {
        // Depend on bundled mb-genres.json being parsed correctly.
        assert_eq!(canonical_genre_name("Jazz"), Some("Jazz".to_string()));
        assert_eq!(canonical_genre_name("JAZZ"), Some("Jazz".to_string()));
        assert_eq!(canonical_genre_name("Hip-Hop"), Some("Hip Hop".to_string()));
        assert_eq!(canonical_genre_name("hip hop"), Some("Hip Hop".to_string()));
    }

    #[test]
    fn canonical_applies_acronym_overrides() {
        assert_eq!(canonical_genre_name("edm"), Some("EDM".to_string()));
        assert_eq!(canonical_genre_name("EDM"), Some("EDM".to_string()));
        assert_eq!(canonical_genre_name("k-pop"), Some("K-Pop".to_string()));
        assert_eq!(canonical_genre_name("K Pop"), Some("K-Pop".to_string()));
        assert_eq!(canonical_genre_name("R&B"), Some("R&B".to_string()));
        assert_eq!(canonical_genre_name("RnB"), Some("R&B".to_string()));
    }

    #[test]
    fn canonical_preserves_unknown_genres() {
        // Made-up genre — should pass through with user's casing intact.
        let unknown = "VaporTwitch";
        assert_eq!(canonical_genre_name(unknown), Some(unknown.to_string()));
    }

    #[test]
    fn canonical_handles_empty_input() {
        assert_eq!(canonical_genre_name(""), None);
        assert_eq!(canonical_genre_name("   "), None);
    }
}
